In [3]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Reading pkl file

In [4]:
df = pd.read_pickle(r'C:\Users\rekha\OneDrive\Documents\Instacart Bascket Analysis\2. Data\prepared Data\orders_products_merged.pkl')

In [5]:
path = r'C:\Users\rekha\OneDrive\Documents\Instacart Bascket Analysis'

In [6]:
path

'C:\\Users\\rekha\\OneDrive\\Documents\\Instacart Bascket Analysis'

In [7]:
df = ords_prods_merged = df = pd.read_pickle(os.path.join(path, '2. Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [8]:
# Creating subset 

df = ords_prods_merged[:1000000]

In [9]:
df

product_id                product_name  aisle_id  department_id  \
0                1  Chocolate Sandwich Cookies        61             19   
1                1  Chocolate Sandwich Cookies        61             19   
2                1  Chocolate Sandwich Cookies        61             19   
3                1  Chocolate Sandwich Cookies        61             19   
4                1  Chocolate Sandwich Cookies        61             19   
...            ...                         ...       ...            ...   
999995        2078  Unsalted Pure Irish Butter        36             16   
999996        2078  Unsalted Pure Irish Butter        36             16   
999997        2078  Unsalted Pure Irish Butter        36             16   
999998        2078  Unsalted Pure Irish Butter        36             16   
999999        2078  Unsalted Pure Irish Butter        36             16   

        prices  order_id  user_id  order_number  orders_day_of_week  \
0          5.8   3139998      138            28                   6   
1          5.8   1977647      138            30                   6   
2          5.8    389851      709             2                   0   
3          5.8    652770      764             1                   3   
4          5.8   1813452      764             3                   4   
...        ...       ...      ...           ...                 ...   
999995    13.5   2280887   102527            25                   6   
999996    13.5   3310791   102530             1                   4   
999997    13.5   1704867   102530             2                   2   
999998    13.5    629818   102530             3                   6   
999999    13.5   2328893   102544             5                   5   

        order_hour_of_day  days_since_prior_order  add_to_cart_order  \
0                      11                     3.0                  5   
1                      17                    20.0                  1   
2                      21                     6.0                 20   
3                      13                     NaN                 10   
4                      17                     9.0                 11   
...                   ...                     ...                ...   
999995                 14                     6.0                  1   
999996                 17                     NaN                 13   
999997                 17                    30.0                  4   
999998                 18                     4.0                  3   
999999                  9                     1.0                  1   

        reordered _merge  
0               0   both  
1               1   both  
2               0   both  
3               0   both  
4               1   both  
...           ...    ...  
999995          1   both  
999996          0   both  
999997          1   both  
999998          1   both  
999999          0   both  

[1000000 rows x 14 columns]

In [10]:
df.shape

(1000000, 14)

In [11]:
# define a function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [12]:
# Aplly the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\rekha\AppData\Local\Temp\ipykernel_10548\3495599497.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [13]:
# fequeancy of price_range 

df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [14]:
# Checking Max value or most expencieve price

df['prices'].max()

24.5

In [15]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\rekha\AppData\Local\Temp\ipykernel_10548\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [16]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [19]:
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High_range_product'

In [20]:
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid_range_product'

In [21]:
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low_range_product'

In [22]:
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid_range_product     21860860
Low_range_product     10126321
High_range_product      417678
Name: count, dtype: int64

In [23]:
ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [24]:
result = []

for value in ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [26]:
ords_prods_merged['Busiest_day'] = result

In [27]:
ords_prods_merged['Busiest_day'].value_counts(dropna = False)

Busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [28]:
ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

# Task 4.7

# 2.Que

In [29]:
# Creating Busiest_days & Checking 2 Slowest_days 


result = []

for value in ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value == 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [30]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

# 3.Que

In [31]:
# Creating Busiest_days column 

ords_prods_merged['Busiest_days'] = result

In [32]:
# Checking Busiest_days column frequency 

ords_prods_merged['Busiest_days'].value_counts(dropna = False)

Busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [33]:
ords_prods_merged.shape

(32404859, 17)

Both of the columns have the same total count of 32,404,859 rows meaning all rows have been allocated

# 4.Que

In [34]:
ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     Busiest_day    Busiest_days  
0  Mid_range_product  Regularly busy  Regularly busy  
1  Mid_range_product  Regularly busy  Regularly busy  
2  Mid_range_product     Busiest day    Busiest days  
3  Mid_range_product  Regularly busy    Slowest days  
4  Mid_range_product      Least busy    Slowest days

In [35]:
# getting value count of each hour 

ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [36]:
# dividing the hours of the day into 3 sections starting with the most orders


result = []

for value in ords_prods_merged["order_hour_of_day"]:
  if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
    result.append("Most_orders")
  elif value == 17 or value == 8 or value == 18 or value == 19 or value == 20 or value == 7 or value == 21 or value == 22:
    result.append("Averagr_orders")
  else:
    result.append("Fewest_orders")

In [37]:
result

['Most_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Most_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Fewest_orders',
 'Averagr_orders',
 'Fewest_orders',
 'Fewest_orders',
 'Fewest_orders',
 'Fewest_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Fewest_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Most_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Averagr_orders',
 'Averagr_orders',
 'Most_orders',
 'Most_ord

In [38]:
# Creating busiest_period_of_day column

ords_prods_merged['Busiest_period_of_day'] = result

# 5.Que

In [39]:
# frequency of new column 

ords_prods_merged['Busiest_period_of_day'].value_counts(dropna = False)

Busiest_period_of_day
Most_orders       21118071
Averagr_orders     9997651
Fewest_orders      1289137
Name: count, dtype: int64

# 7. Que

In [40]:
# Exporing pickle file

In [41]:
ords_prods_merged.to_pickle(os.path.join(path, '2. Data', 'Prepared Data', 'orders_products_merged_New.pkl'))

In [42]:
ords_prods_merged_New.head()

NameError: name 'ords_prods_merged_New' is not defined